packages/core/src/shared/languageServer/lspResolver.ts

```diff
@@ -0,0 +1,352 @@
+        const compatibleLspVersions = this.compatibleManifestLspVersion()
+
+        // determine all folders containing lsp versions in the fallback parent folder
+        const cachedVersions = (await fs.readdir(this.defaultDownloadFolder()))
+            .filter(([_, filetype]) => filetype === FileType.Directory)
+            .map(([pathName, _]) => semver.parse(pathName))
+            .filter((ver): ver is semver.SemVer => ver !== null)
+
+        const expectedVersion = semver.parse(version)
+        if (!expectedVersion) {
+            return undefined
+        }
+
+        const sortedCachedLspVersions = compatibleLspVersions
+            .filter((v) => this.isValidCachedVersion(v, cachedVersions, expectedVersion))
+            .sort((a, b) => semver.compare(a.serverVersion, b.serverVersion))
+
+        const fallbackDir = (
+            await Promise.all(sortedCachedLspVersions.map((ver) => this.getValidLocalCacheDirectory(ver)))
+        ).filter((v) => v !== undefined)
+        return fallbackDir.length > 0 ? fallbackDir[0] : undefined
+    }
+
+    /**
+     * Validate the local cache directory of the given lsp version (matches expected hash)
+     * If valid return cache directory, else return undefined
+     */
+    private async getValidLocalCacheDirectory(version: LspVersion) {
+        const targetContents = this.getTargetContents(version)
+        if (targetContents === undefined || targetContents.length === 0) {
+            return undefined
+        }
+
+        const cacheDir = this.getDownloadDirectory(version.serverVersion)
+        const hasValidCache = await this.hasValidLocalCache(cacheDir, targetContents)
+
+        return hasValidCache ? cacheDir : undefined
+    }
+
+    /**
+     * Determines if a cached LSP version is valid for use as a fallback.
+     * A version is considered valid if it exists in the cache and is less than
+     * or equal to the expected version.
+     */
+    private isValidCachedVersion(version: LspVersion, cachedVersions: semver.SemVer[], expectedVersion: semver.SemVer) {
+        const serverVersion = semver.parse(version.serverVersion) as semver.SemVer
+        return cachedVersions.find((x) => x === serverVersion) && serverVersion <= expectedVersion
+    }
+
+    /**
+     * Download and unzip all of the contents into the download directory
+     *
+     * @returns
+     *  true, if all of the contents were successfully downloaded and unzipped
+     *  false, if any of the contents failed to download or unzip
+     */
+    private async downloadRemoteTargetContent(contents: TargetContent[], version: string) {
+        const downloadDirectory = this.getDownloadDirectory(version)
+
+        if (!(await fs.existsDir(downloadDirectory))) {
+            await fs.mkdir(downloadDirectory)
+        }
+
+        const downloadTasks = contents.map(async (content) => {
+            // TODO This should be using the retryable http library but it doesn't seem to support zips right now
```

Comment:
Calling this out as well, we should switch to something that has retries with backoffs